File: automata/src/main/java/automata/sim/automata/State.java
package automata.sim.automata;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class State {
    private String name;
    private boolean isAcceptState;
    private Map<Character, Set<State>> transitions;

    public State(String name, Boolean isAcceptState) {
        this.name = name;
        this.isAcceptState = isAcceptState;
        this.transitions = new HashMap<>();
    }

    public void addTransition(char input, State state) {
        this.transitions.computeIfAbsent(input, k -> new HashSet<>()).add(state);
    }

    public void removeTransition(char input, State state) {
        this.transitions.remove(input, state);
    }

    public Set<State> getNextState(char input) { // valid states from given input
        return this.transitions.getOrDefault(input, new HashSet<>());
    }

    public boolean isAcceptState() {
        return isAcceptState;
    }
}

File: automata/AutomatonBackend/State.cs
using System.Collections.Generic;

namespace AutomatonBackend.Controllers;


    public class State
    {
        public string Name { get; private set; }
        public bool IsAcceptState { get; private set; }
        public Dictionary<char, HashSet<State>> transitions;

        public State(string name, bool isAcceptState)
        {
            Name = name;
            IsAcceptState = isAcceptState;
            this.transitions = new Dictionary<char, HashSet<State>>();
        }

        public void AddTransition(char input, State state)
        {
            if (!this.transitions.ContainsKey(input))
            {
                this.transitions[input] = new HashSet<State>();
            }
            this.transitions[input].Add(state);
        }

        public void RemoveTransition(char input, State state)
        {
            if (this.transitions.ContainsKey(input))
            {
                this.transitions[input].Remove(state);
            }
        }

        public HashSet<State> GetNextState(char input)
        {            
            return this.transitions.GetValueOrDefault(input, new HashSet<State>());
        }

      
    public void RemoveStateFromTransitions(State stateToRemove)
    {
        foreach (var charKvp in transitions)
        {
            char input = charKvp.Key;
            HashSet<State> statesHashSet = charKvp.Value;

            statesHashSet.Remove(stateToRemove);

            // Remove the entire char key if the HashSet<State> becomes empty
            if (statesHashSet.Count == 0)
            {
                transitions.Remove(input);
            }
        }
    }
    }


